File: cpp/leetcode/285.cpp
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        if(p->right){
            TreeNode* node = p->right;
            while(node->left) node = node->left;
            return node;
        }
        TreeNode *node = NULL;
        while(root){
            if(root->val > p->val){
                node = root;
                root = root->left;
            } else if (root->val < p->val) {
                root = root->right;
            } else return node;
        }
        return node;
    }
};

File: python/leetcode/285.py
class Solution:
    def inorderSuccessor(self, root, p):
        if p.right:
            node = p.right
            while node.left:
                node = node.ledt
            return node

        res = None
        while root:
            if p.val < root.val:
                res = root
                root = root.left
            elif p.val > root.val:
                root = root.right
            else:
                break
        return res
